debian/speechd.7
.\" Copyright (C) 2003 Brailcom, o.p.s.
.\"
.\" This is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2, or (at your option) any later
.\" version.
.\" 
.\" This is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\" 
.\" You should have received a copy of the GNU General Public License with
.\" your Debian GNU/Linux system, in /usr/share/common-licenses/GPL, or with
.\" the dpkg source package as the file COPYING.  If not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH SPEECHD 7 
.SH NAME
Speech Dispatcher
.SH DESCRIPTION
Speech Dispatcher does not have a useful manpage.  Please look into its Info
documentation for information about any program within the Speech Dispatcher
packages.
.SH SEE ALSO
.BR info (1)


NEWS
Version 0.12
* Add openjtalk module.
* Add socket activation.
* Add language and variant parameters to "list voice" command.
* libspeechd: Simplify buffer management.
* Fix use of semaphores on MacOS.
* Fix splitting long UTF-8 sequences.
* Update CLDR to version 44 and symbols from NVDA.
* Add spd_fd function to C api.

Version 0.11.5
* Update CLDR to version 43 and symbols from NVDA.
* Fix parsing and emitting space character.
* Reduce espeak buffer size to avoid ssml issues.
* Add --disable-doc.
* Fix spd-conf not being able to set the default module.
* Fix pausing without index.
* ibmtts: Fix parameter reset.
* ibmtts: Disable using the module punctuation by default.

Version 0.11.4
* Update CLDR to version 42 and symbols from NVDA.
* Fix audio plugin loading with dlopen.
* Fix atomicity of getting reply in threaded mode.

Version 0.11.3
* Fix back DefaultModule configuration.
* pico: Avoid falling to english when passed a bogus voice name.
* espeak: Fix setting voice type.

Version 0.11.2
* Fix loading xx-yy locales.
* Various memory leaks fixes.
* Add mimic3 configuration file.
* pico: Fix setting language vs voice.
* Make sure that modules report a list of voices.
* Update CLDR to version 41, symbols from NVDA and orca.
* Allow building without ltdl.
* Re-enable SSML in espeak-ng-mbrola module.

Version 0.11.1
* Add SPEECHD_PLUGIN_DIR environment variable.
* Fix listing voices of the default module.

Version 0.11
* Support playing audio through the server.
* Simplify modules thanks to this.
* modules: Add support for loading from user's .local/libexec/speech-dispatcher.
* symbols: Process symbols.dic before emojis.dic.
* symbols: Enable speechd symbols processing by default.
* modules: Moved speech dispatcher modules to /usr/libexec/speech-dispatcher-modules
  (but keep compatibility with old /usr/lib/speech-dispatcher-modules path).
* espeak-ng: Add support for mbrola voices.
* ibmtts/voxin: Improve language selection.
* ibmtts/voxin: Add option to disable punctuation commands.
* ibmtts: Fix library load.
* ibmtts: Do not build by default on other systems than 32bit PC.
* mary: Add auto-detection.
* mary: Add newer voices.
* mary: Add volume, pitch, and rate support.
* ivona: Add auto-detection.
* festival: Strip head silence.
* generic: Add DefaultVoice option.
* es_ES: Add some gender neutral rules.
* Add SPEECHD_CMD environment variable.
* modules: Rewrite main functions with BSD licence, to let proprietary modules easily reuse this as a basis.
* modules: Add skeletons ready for use as a basis for new modules.
* Add script to run speechd from the build tree.
* Update CLDR to version 39, symbols from NVDA and orca.
* Add Esperanto translation.
* Sort modules by quality, let the best quality module be the default.

Version 0.10.2
* generic: Add support for sound icons
* Add French, Norwegian Bokmål, Portuguese, Brazilian Portuguese, and Russian translations.
* Add voxin module variant.
* Factorize espeak module with espeak-ng module.
* Make ibmtts and voxin use the common queue helpers.
* Fix python bindings against python 3.9.
* Reference the golang bindings.
* Add spd_get_client_id.
* Make key, char and sound_icon commands return message id.
* Expose modulebindir in speech-dispatcher.pc.
* Remove generic modules for synths that have a non-generic module.

Version 0.10
* Add punctuation level "most".
* Rework thread synchronization.
* Move espeak/espeak-ng audio queuing to shared file, rework it.
* Make baratinoo use this audio queueing, enable marks, and support
responsiveness parameter.
* Add support for Baratinoo engine 8.6
* Show mark progression in spd-say.
* Add --character option to spd-say.
* Put country/region in language instead of variant.
* pico, generic: fix setting voice by language.
* generic: Add fallback voices for generic languages.
* symbols: Add support for SSML.
* symbols: Add SymbolPreprocFile directive to load several symbol files.
* symbols: Replace DefaultSymbolsPreproc directive with SymbolsPreproc directive
to have more fine-grained control over server preprocessing.
* symbols: Import emojis and unicode font variants support from NVDA and Unicode
CLDR and UnicodeData.
* symbols: Import some symbols from Orca.
* symbols: Add rules to drop gender-neutral forms.
* symbols: Add support for group references.
* Disable Mary-TTS module by default.
* Replace AudioPulseServer option with AudioPulseDevice.
* Fix default pulse latency to 10ms to avoid pulseaudio distorted output.
* Made client isolation stronger for Snap/Flatpak support.

Version 0.9.1
* Add module for the non-free IBM TTS (voxin) speech synthesis.
* Extend licence to later versions of GPL and LGPL.
* Update mailing list address to savannah.
* Make generic modules fallback to existing voices.

Version 0.9
* Add modules for non-free Baratinoo (VoxyGen) and Kali speech syntheses.
* Add configuration file for the Mary-TTS system.
* Add configuration file for espeak-ng + mbrola.
* Set the pulse client name when using the generic module with paplay.
* espeak-*-mbrola-generic: Update voice list.
* Auto-detect module availability.
* Make generic module provide voice list.
* Add systemd service file.

Version 0.8.8
* Add German translation, thanks to Chris Leick for the patch
* Fix some spelling mistakes, thanks to Paul Gevers for the patch
* Some slight code improvements

Version 0.8.7
* Further fixes to spd-conf, which should now work properly.
* Split the espeak-ng driver code into its own source file.
* Add a work-around to the espeak-ng driver to account for spaces in voice
  names which recently appeared in espeak-ng git master. This will properly
  be fixed in 0.9.
* Voice names are not forced to lower case, due to espeak-ng git master now
  having multi-case voice names.
* Fix stripped audio output from the flite module, thanks to Samuel Thibault.
* Further code and build improvements.

Version 0.8.6
* Various internal code improvements.
* Fix more compiler warnings.

src/modules/python/src/pySpeechModule/speechd.py
# MIT License

# Copyright (c) 2026 John Settlemyer

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import logging
import time
import soundfile as sf

class SpeechDispatch():
    def __init__(self):
        """
        you should place the voices list here in __init__

        .. code-block:: python

            self.voices = [{"name": "Testing", "language": "en", "variant": "none"},
            {"name": "Testing2", "language": "en", "variant": "none"}]

        You should also set your max_ahead and min_ahead here.

        .. code-block:: python

            self.max_ahead = 60 # this sets the max number of second your generation will get ahead of played audio.
            self.min_ahead = 30 # this sets how small the ahead number maybe before starting generation again.

        You can also place model init or what ever else you would like to init here.
        """
        pass
    def _set_init(self,queue, server):
        """
        Set the action_queue and server.

        :meta private:
        """

        self.action_queue = queue
        self._server = server
    def configure(self, configure_file: str):
        """
        This method get passed in the file path to the modules dot conf file
        If you need to use it you should open the file and parse here.
        """
        pass
    def change_voice(self, voice: str):
        """
        This method gets passed the name of a voice when the user has requested to change the voice.

        :param voice: String containing the voice name.
        :type String: String
        """
        pass
    def change_language(self, language: str):
        """
        This method gets passed a language when the user has requested to change the language.

        :param voice: String containing the language.
        :type String: String
        """
        pass
    def change_speed(self, speed: str):
        """
        This method gets passed a speed when the user requests to change the speed.
        The speed will be between -100 and 100

        :param voice: String containing the new speed.
        :type String: String
        """
        pass
    def change_pitch(self, pitch: str):
        """
        This method gets passed a pitch when the user requests to change the pitch

        :param voice: String containing the new pitch.
        :type String: String
        """
        pass
    def _parse_settings(self,options: dict):
        """
        When passed the settings it calls the convenience methods associated with that setting.

        :meta private:
        """
        if 'synthesis_voice' in options:
            self.change_voice(options['synthesis_voice'])
        if 'language' in options:
            self.change_language(options['language'])
        if 'rate' in options:
            self.change_speed(options['rate'])
        if 'pitch' in options:
            self.change_pitch(options['pitch'])
    def _settings(self,options):
        """
        Passed the setting, sends it to the parsers.

        :meta private:
        """
        self._parse_settings(options)
        self.settings(options)

    def settings(self,options: dict):
        """
        This gets passed the full settings options

        :param options: Dict containing all the settings options that were sent.
        :type Dictionary: Dictionary.
        """
        pass
    def _speak(self,text: str):
        """
        This is called to generate speech
        It then iterate's through speech chunks returned by the user.

        :meta private:
        """
        self._server.stop_set(False)
        self._ahead = 0
        self.action_queue.put({"command": "begin"})
        it = self.speak(text)
        while (True):
            try:
                if (self._server.stop_get()):
                    break
                start_time = time.time()
                item = next(it) # get the next audio chunk
                end_time = time.time()
                total_time = end_time-start_time # total time it took to generate that audio chunk
                produced_time = (len(item['audio']) / item['rate']) # amount of audio generate
                logging.debug(f"Speak chunk took {total_time} to produce {produced_time} for mark {item['mark']}")
                self._ahead = self._ahead + produced_time
                self._ahead = self._ahead - total_time
                logging.debug(f"Ahead: {self._ahead}")
                self.action_queue.put({"command": "speak", "val": item})
                # sleep if we have too much audio generated, only do this is max_ahead and min_ahead are set.
                if (hasattr(self, "max_ahead") and hasattr(self, "min_ahead")):
                    if (self._ahead >= self.max_ahead):
                        logging.debug(f"Ahead by too much sleeping")
                        while (self._ahead >= self.min_ahead):
                            time.sleep(1)
                            self._ahead = self._ahead - 1
                            if (self._server.stop_get()):
                                break;
            except StopIteration:
                logging.debug("finished speak chunking")
                break
        self.action_queue.put({"command": "end"})
    def speak(self,text: str):
        """
        This method get called to get the user to generate audio.
        Data must be yield'ed in the format

        .. code-block:: python

            yield {"audio": <data>, "rate": <samplerate>, "mark": <mark>}

        :param text: String containing ssml text.
        :type String: String
        :yields: Dict containing keys for "audio", "rate", "mark"
        :ytype: Dict
        """
        pass